crates/starknet-devnet/tests/test_blocks_on_demand.rs

```diff
@@ -0,0 +1,185 @@
+        assert_tx_successful, get_contract_balance,
+        get_simple_contract_in_sierra_and_compiled_class_hash,
+    };
+
+    static DUMMY_ADDRESS: u128 = 1;
+    static DUMMY_AMOUNT: u128 = 1;
+
+    async fn assert_latest_block_with_transactions(
+        devnet: &BackgroundDevnet,
+        block_number: u64,
+        transactions: Vec<FieldElement>,
+    ) {
+        let latest_block = devnet.get_latest_block_with_tx_hashes().await.unwrap();
+        assert_eq!(latest_block.block_number, block_number);
+        assert_eq!(transactions, latest_block.transactions);
+        assert_eq!(latest_block.status, BlockStatus::AcceptedOnL2);
+
+        for tx_hash in latest_block.transactions {
+            assert_tx_successful(&tx_hash, &devnet.json_rpc_client).await;
+        }
+    }
+
+    async fn assert_pending_block_with_transactions(
+        devnet: &BackgroundDevnet,
+        block_number: u64,
+        transactions_count: u128,
+    ) {
+        let pending_block = devnet.get_pending_block_with_tx_hashes().await.unwrap();
+        assert_eq!(pending_block.block_number, block_number);
+        assert_eq!(pending_block.transactions.len() as u128, transactions_count);
+        assert_eq!(pending_block.status, BlockStatus::Pending);
+    }
+
+    async fn assert_balance(devnet: &BackgroundDevnet, expected: FieldElement, tag: BlockTag) {
+        let balance = devnet
+            .get_balance_by_tag(
+                &FieldElement::from_hex_be(DUMMY_ADDRESS.to_string().as_str()).unwrap(),
+                FeeUnit::WEI,
+                tag,
+            )
+            .await
+            .unwrap();
+        assert_eq!(balance, expected);
+    }
+
+    #[tokio::test]
+    async fn blocks_on_demand_states_and_blocks() {
+        let devnet =
+            BackgroundDevnet::spawn_with_additional_args(&["--blocks-on-demand"]).await.unwrap();
+
+        let tx_count = 5;
+        let mut tx_hashes = Vec::new();
+        for _ in 0..tx_count {
+            let mint_hash = devnet.mint(DUMMY_ADDRESS, DUMMY_AMOUNT).await;
+            tx_hashes.push(mint_hash);
+        }
+
+        assert_balance(&devnet, FieldElement::from(tx_count * DUMMY_AMOUNT), BlockTag::Pending)
+            .await;
+        assert_balance(&devnet, FieldElement::from(0_u128), BlockTag::Latest).await;
+
+        devnet.create_block().await.unwrap();
+
+        assert_balance(&devnet, FieldElement::from(tx_count * DUMMY_AMOUNT), BlockTag::Pending)
+            .await;
+        assert_balance(&devnet, FieldElement::from(tx_count * DUMMY_AMOUNT), BlockTag::Latest)
+            .await;
+
+        assert_latest_block_with_transactions(&devnet, 1, tx_hashes).await;
+
+        // check if pending_block was restarted
+        let pending_block = devnet.get_pending_block_with_tx_hashes().await.unwrap();
+        assert_eq!(pending_block.block_number, 2);
+        assert_eq!(pending_block.transactions.len(), 0);
+
+        assert_pending_block_with_transactions(&devnet, 2, 0).await;
```

Comment:
Why are these two assertions done both here and in `assert_pending_block_with_transactions`? If this is the only place where the function is used, it's ok to remove it. Also, perhaps better to assert tx hash than tx count

Reply:
Right, good that you spotted that, somehow I missed that. My mistake, fixed in b8fc6af3cac957f64a722c0a872c57e473a4fa18